Add support for stomp over tls/ssl.

kgoerlitz opened the pull request:
This implements request #6 (Need to support stomp protocol over SSL/TLS connection).

kgoerlitz committed:
Add support for stomp over tls/ssl.

kgoerlitz committed:
Merge branch 'master' of https://github.com/logstash-plugins/logstash-input-stomp into master

kgoerlitz committed:
Merge branch 'master' of https://github.com/logstash-plugins/logstash-input-stomp

kgoerlitz committed:
Merge branch 'master' of https://github.com/kgoerlitz/logstash-input-stomp.git

jordansissel commented on lib/logstash/inputs/stomp.rb:
```diff
@@ -37,6 +37,12 @@ class LogStash::Inputs::Stomp < LogStash::Inputs::Base
+  config :ssl, :validate => :boolean, :default => false
+
+  # Validate TLS/SSL certificate?
+  config :verify_ssl, :validate => :boolean, :default => false
```
Two comments:

* Do we need this setting? Disabling SSL verification means you are turning off the security provided by SSL, and why are we using SSL in this case if we don't want security?

* This should default to true, not false.

kgoerlitz replied:
Disabling SSL certificate verification only disables the checks that the certificate is specifically for the target host and that is signed by a good certificate authority.  The setting is needed to support self-signed certificates.

I agree that the default should probably be true.  I will make that change.

jordansissel commented on lib/logstash/inputs/stomp.rb:
```diff
@@ -37,6 +37,12 @@ class LogStash::Inputs::Stomp < LogStash::Inputs::Base
   config :reconnect_interval, :validate => :number, :default => 30
 
+  # Enable TLS/SSL connection?
+  config :ssl, :validate => :boolean, :default => false
```
This leaves no easy way for users to provide trusted certificate authorities. Do you think you could add that?

kgoerlitz replied:
Trusted certificate authorities are supported by the underlying Java cacerts file.  Overriding the standard truststore, if needed, can be accomplished by setting the Java system property javax.net.ssl.trustStore.

jordansissel replied:
This is a hard thing for users to do correctly and it also puts the configuration in too many places. Because of this, we try to offer a setting for plugins using SSL that users can specify a file (openssl or keystore format, if both are possible):

* [elasticsearch output `cacert`](https://www.elastic.co/guide/en/logstash/current/plugins-outputs-elasticsearch.html#plugins-outputs-elasticsearch-cacert)
* [beats input `ssl_certificate_authorities`](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-beats.html#plugins-inputs-beats-ssl_certificate_authorities)
* [http output `cacert`](https://www.elastic.co/guide/en/logstash/current/plugins-outputs-http.html#plugins-outputs-http-cacert)

We aren't consistent in naming, unfortunately, but we are trying to provide this setting explicitly for users.

jordansissel replied:
I understnad if you aren't willing or interested in adding explicit certificate support. Maybe we can compromise.

Can you instead add documentation that tells users the necessary steps for creating a keystore with their trusted certificate authorities and also how to provide this file to Logstash, and also the caveats (it overrides the global keystore used by the whole Java process)?

kgoerlitz replied:
I added support for specification of cacert, client_cert, and client_key, changed ssl_verify to ssl_certificate_validation. to match other plugins. I also updated the doc.

jordansissel reviewed (changes requested):
Thank you for your contribution! I added some comments to the patch.

jordansissel commented:
New features/documentation also now lives in docs/index.asciidoc

kgoerlitz committed:
Set the default for verify_ssl to true.

jordansissel commented:
>  The setting is needed to support self-signed certificates.

You can add self-signed certs to your trust store. If you don't do this, and instead disable verification, you cannot prevent your client (stomp client) from talking to an untrusted server (whether through misconfiguration or through MITM) because you trust all servers.

kgoerlitz committed:
Add support for specification of cacert, client_cert, and client_key.  Change ssl_verify to ssl_certificate_validation.  Update doc.

kgoerlitz commented:
I added support for specification of cacert, client_cert, and client_key, changed ssl_verify to ssl_certificate_validation. to match other plugins.  I also updated the doc.

kgoerlitz commented:
> You can add self-signed certs to your trust store.

Yes, but that can be a pain if you are building up dev and QA systems on a daily basis, which we are.

kgoerlitz commented:
Do you require any more changes?

tetebueno commented:
Hi, any news on this? It's just what we're looking for.
Cheers.

kgoerlitz committed:
Merge branch 'master' of https://github.com/logstash-plugins/logstash-input-stomp.git

tetebueno commented:
It seems requested changes were made. Is there anything left needed for this to be merged?

Sivrat commented:
The changes made back in 2017 appear to work fine. 
Can someone actually finish merging the work completed 3 years ago? Or are more changes needed before people can use secure communications for STOMP servers with Logstash?
@jordansissel ?
